File: frontend/src/engine/domains.ts
/** Offline domain and URL heuristics. Mirrors app/checks/domains.py. */

import {
  COMPANY_STOPWORDS,
  FREE_EMAIL_PROVIDERS,
  KNOWN_BRANDS,
  MULTI_PART_SUFFIXES,
  SUSPICIOUS_TLDS,
  URL_SHORTENERS,
  type Brand,
} from "./data";

const HOMOGLYPHS: Record<string, string> = { "0": "o", "1": "l", "3": "e", "4": "a", "5": "s", "7": "t", $: "s" };
const SHORT_TOKEN_LEN = 5;
const PRODUCT_MENTION_RE =
  /\b(?:google\s+(?:meet|pay|play|forms?|drive|docs|maps|chrome)|gmail|microsoft\s+(?:teams|word|excel|office|forms)|amazon\s+pay|paytm\s+(?:app|wallet)|jio\s*(?:meet|sim))\b/gi;
const IPV4_RE = /^\d{1,3}(?:\.\d{1,3}){3}$/;

/** Lower-case host of a URL; scheme-less input is allowed. */
export function hostOf(url: string): string | null {
  const candidate = url.includes("://") ? url : `http://${url}`;
  try {
    const host = new URL(candidate).hostname;
    return host ? host.replace(/\.+$/, "").toLowerCase() : null;
  } catch {
    return null;
  }
}

/** `a.b.tcs.com` -> `tcs.com`; null for IPs and single-label hosts. */
export function registrableDomain(host: string): string | null {
  const clean = host.replace(/\.+$/, "").toLowerCase();
  if (IPV4_RE.test(clean) || clean.includes(":")) return null;
  const labels = clean.split(".");
  if (labels.length < 2) return null;
  const take = MULTI_PART_SUFFIXES.has(labels.slice(-2).join(".")) && labels.length >= 3 ? 3 : 2;
  return labels.slice(-take).join(".");
}

const tldOf = (domain: string): string => domain.split(".").pop() ?? domain;
export const emailDomain = (email: string): string => (email.split("@").pop() ?? "").toLowerCase();
const labelOf = (domain: string): string => domain.split(".")[0];

export function levenshtein(a: string, b: string): number {
  if (a.length < b.length) [a, b] = [b, a];
  let previous = Array.from({ length: b.length + 1 }, (_, i) => i);
  for (let i = 1; i <= a.length; i++) {
    const current = [i];
    for (let j = 1; j <= b.length; j++) {
      current.push(Math.min(previous[j] + 1, current[j - 1] + 1, previous[j - 1] + (a[i - 1] === b[j - 1] ? 0 : 1)));
    }
    previous = current;
  }
  return previous[previous.length - 1];
}

export const isOfficialDomain = (domain: string, brand: Brand): boolean =>
  brand.domains.some((d) => domain === d || domain.endsWith(`.${d}`));

const isOfficialAnywhere = (domain: string): boolean => KNOWN_BRANDS.some((b) => isOfficialDomain(domain, b));

function imitates(label: string, token: string): boolean {
  if (label.split("-").includes(token)) return true;
  const compact = label.replace(/-/g, "");
  if (token.length > SHORT_TOKEN_LEN && compact.includes(token)) return true;
  const deglyphed = [...compact].map((c) => HOMOGLYPHS[c] ?? c).join("");
  if (deglyphed === token || (token.length > SHORT_TOKEN_LEN && deglyphed.includes(token))) return true;
  if (token.length >= SHORT_TOKEN_LEN) {
    const allowed = token.length >= 8 ? 2 : 1;
    // Edit distance is at least the length gap: skip the O(n*m) DP when it can't match.
    if (Math.abs(compact.length - token.length) > allowed) return false;
    const distance = levenshtein(compact, token);
    return distance > 0 && distance <= allowed;
  }
  return false;
}

export function findLookalikeDomains(domains: string[]): string[] {
  const hits: string[] = [];
  for (const domain of domains) {
    if (isOfficialAnywhere(domain) || FREE_EMAIL_PROVIDERS.has(domain)) continue;
    const label = labelOf(domain);
    const brand = KNOWN_BRANDS.find((b) => b.tokens.some((t) => imitates(label, t)));
    if (brand) hits.push(`${domain} imitates ${brand.name} (official: ${brand.domains.join(", ")})`);
  }
  return hits;
}

export const findFreeEmailSenders = (emails: string[]): string[] =>
  emails
    .filter((e) => FREE_EMAIL_PROVIDERS.has(emailDomain(e)))
    .map((e) => `${e} uses free provider ${emailDomain(e)}`);

const escapeRegExp = (value: string): string => value.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");

// One whole-word pattern per brand, compiled once at load.
const BRAND_ALIAS_RES: [Brand, RegExp][] = KNOWN_BRANDS.map((b) => [
  b,
  new RegExp(`\\b(?:${b.aliases.map(escapeRegExp).join("|")})\\b`),
]);

function brandsMentioned(text: string): Brand[] {
  const lowered = text.toLowerCase();
  return BRAND_ALIAS_RES.filter(([, re]) => re.test(lowered)).map(([b]) => b);
}

function claimedBrands(companyName: string | null, text: string): Brand[] {
  if (companyName) return brandsMentioned(companyName);
  return brandsMentioned(text.replace(PRODUCT_MENTION_RE, " "));
}

function nameMatchesDomain(company: string, domain: string): boolean {
  const label = labelOf(domain).replace(/-/g, "");
  const allWords = company.toLowerCase().match(/[a-z0-9]+/g) ?? [];
  if (allWords.some((w) => !COMPANY_STOPWORDS.has(w) && w.length >= 3 && label.includes(w))) return true;
  const acronym = allWords
    .filter((w) => !["pvt", "ltd", "private", "limited"].includes(w))
    .map((w) => w[0])
    .join("");
  return acronym.length >= 2 && label.includes(acronym);
}

export function findCompanyDomainMismatch(companyName: string | null, text: string, senderEmails: string[]): string[] {
  const senderDomains = [
    ...new Set(senderEmails.map((e) => registrableDomain(emailDomain(e)) ?? emailDomain(e))),
  ];
  if (senderDomains.length === 0) return [];
  const hits: string[] = [];
  const brands = claimedBrands(companyName, text);
  for (const brand of brands) {
    if (!senderDomains.some((d) => isOfficialDomain(d, brand))) {
      hits.push(`Letter mentions ${brand.name} but emails come from ${senderDomains.join(", ")}`);
    }
  }
  if (companyName && brands.length === 0 && !senderDomains.some((d) => nameMatchesDomain(companyName, d))) {
    hits.push(`Company '${companyName}' does not match sender domain(s) ${senderDomains.join(", ")}`);
  }
  return hits;
}

export const findSuspiciousTlds = (domains: string[]): string[] =>
  domains.filter((d) => SUSPICIOUS_TLDS.has(tldOf(d))).map((d) => `${d} (.${tldOf(d)})`);

export const findShorteners = (urls: string[]): string[] => urls.filter((u) => URL_SHORTENERS.has(hostOf(u) ?? ""));

export const findInsecureUrls = (urls: string[]): string[] => urls.filter((u) => u.toLowerCase().startsWith("http://"));

File: app/checks/domains.py
"""Offline domain and URL heuristics.

Free-email senders, company/domain mismatch, typosquat / look-alike brand
domains, suspicious TLDs, URL shorteners and plain-HTTP links.
"""

import ipaddress
import re
from urllib.parse import urlsplit

from app.checks.data import (
    COMPANY_STOPWORDS,
    FREE_EMAIL_PROVIDERS,
    KNOWN_BRANDS,
    MULTI_PART_SUFFIXES,
    SUSPICIOUS_TLDS,
    URL_SHORTENERS,
    Brand,
)

# Common character substitutions used in look-alike domains.
_HOMOGLYPHS = str.maketrans({"0": "o", "1": "l", "3": "e", "4": "a", "5": "s", "7": "t", "$": "s"})
_SHORT_TOKEN_LEN = 5  # tokens this short must match a whole hyphen-separated part
_PRODUCT_MENTION_RE = re.compile(
    r"\b(?:google\s+(?:meet|pay|play|forms?|drive|docs|maps|chrome)|gmail|"
    r"microsoft\s+(?:teams|word|excel|office|forms)|amazon\s+pay|"
    r"paytm\s+(?:app|wallet)|jio\s*(?:meet|sim))\b",
    re.IGNORECASE,
)
# One precompiled whole-word pattern per brand, built once at import instead of
# compiling a regex per alias on every call.
_BRAND_ALIAS_RES: tuple[tuple[Brand, re.Pattern[str]], ...] = tuple(
    (brand, re.compile(r"\b(?:" + "|".join(re.escape(a) for a in brand.aliases) + r")\b"))
    for brand in KNOWN_BRANDS
)


def host_of(url: str) -> str | None:
    """Return the lower-case host of a URL; scheme-less input is allowed."""
    candidate = url if "://" in url else f"http://{url}"
    try:
        host = urlsplit(candidate).hostname
    except ValueError:
        return None
    return host.rstrip(".").lower() if host else None


def registrable_domain(host: str) -> str | None:
    """Reduce a host to its registrable domain (``a.b.tcs.com`` -> ``tcs.com``).

    Returns None for IP addresses and single-label hosts.
    """
    host = host.rstrip(".").lower()
    try:
        ipaddress.ip_address(host)
        return None
    except ValueError:
        pass
    labels = host.split(".")
    if len(labels) < 2:
        return None
    take = 3 if ".".join(labels[-2:]) in MULTI_PART_SUFFIXES and len(labels) >= 3 else 2
    return ".".join(labels[-take:])


def tld_of(domain: str) -> str:
    """Last label of a domain."""
    return domain.rsplit(".", 1)[-1]


def email_domain(email: str) -> str:
    """Lower-case domain part of an email address."""
    return email.rsplit("@", 1)[-1].lower()


def _label(domain: str) -> str:
    """Registrable domain without its public suffix (``tcs-careers.co.in`` -> ``tcs-careers``)."""
    return domain.split(".", 1)[0]


def levenshtein(a: str, b: str) -> int:
    """Classic edit distance (insert / delete / substitute), O(len(a)*len(b))."""
    if len(a) < len(b):
        a, b = b, a
    previous = list(range(len(b) + 1))
    for i, char_a in enumerate(a, start=1):
        current = [i]
        for j, char_b in enumerate(b, start=1):
            current.append(min(
                previous[j] + 1,
                current[j - 1] + 1,
                previous[j - 1] + (char_a != char_b),
            ))
        previous = current
    return previous[-1]


def is_official_domain(domain: str, brand: Brand) -> bool:
    """True if `domain` is (a subdomain of) one of the brand's official domains."""
    return any(domain == d or domain.endswith(f".{d}") for d in brand.domains)


def _is_official_anywhere(domain: str) -> bool:
    return any(is_official_domain(domain, brand) for brand in KNOWN_BRANDS)


def _imitates(label: str, token: str) -> bool:
    """Does a domain label imitate a brand token?

    Matches (1) the token as a whole hyphen-separated part
    (``tcs-careers``), (2) for longer tokens, the token embedded anywhere
    (``infosyscareers``), (3) homoglyph swaps (``inf0sys``), or
    (4) a small edit distance (``infosis``, ``accentrue``).
    """
    parts = label.split("-")
    if token in parts:
        return True
    compact = label.replace("-", "")
    if len(token) > _SHORT_TOKEN_LEN and token in compact:
        return True
    deglyphed = compact.translate(_HOMOGLYPHS)
    if deglyphed == token or (len(token) > _SHORT_TOKEN_LEN and token in deglyphed):
        return True
    if len(token) >= _SHORT_TOKEN_LEN:
        allowed = 2 if len(token) >= 8 else 1
        # Edit distance is at least the length gap: skip the O(n*m) DP when it can't match.
        if abs(len(compact) - len(token)) > allowed:
            return False
        return 0 < levenshtein(compact, token) <= allowed
    return False


def find_lookalike_domains(domains: list[str]) -> list[str]:
    """Domains that imitate a known employer without belonging to it."""
    hits: list[str] = []
    for domain in domains:
        if _is_official_anywhere(domain) or domain in FREE_EMAIL_PROVIDERS:
            continue
        label = _label(domain)
        for brand in KNOWN_BRANDS:
            if any(_imitates(label, token) for token in brand.tokens):
                hits.append(f"{domain} imitates {brand.name} (official: {', '.join(brand.domains)})")
                break
    return hits


def find_free_email_senders(emails: list[str]) -> list[str]:
    """Emails hosted by a free consumer provider (Gmail, Yahoo, Outlook...)."""
    return [
        f"{email} uses free provider {email_domain(email)}"
        for email in emails
        if email_domain(email) in FREE_EMAIL_PROVIDERS
    ]


def brands_mentioned(text: str) -> list[Brand]:
    """Known brands whose name appears in the text as a whole word."""
    lowered = text.lower()
    return [brand for brand, pattern in _BRAND_ALIAS_RES if pattern.search(lowered)]


def claimed_brands(company_name: str | None, text: str) -> list[Brand]:
    """Known brands the letter claims to come from.

    Prefers Gemini's `company_name`. Without it, scans the text but first
    removes product mentions ("Google Meet", "Amazon Pay") so a genuine letter
    that schedules a Teams call is not treated as claiming to be Microsoft.
    """
    if company_name:
        return brands_mentioned(company_name)
    return brands_mentioned(_PRODUCT_MENTION_RE.sub(" ", text))


def _name_matches_domain(company: str, domain: str) -> bool:
    """Loose match between a company name and a domain label.

    True if a significant word of the name (or its acronym) appears in the
    label, e.g. "Nexora Analytics Pvt Ltd" ~ nexora.in, "Tata Consultancy
    Services" ~ tcs.com.
    """
    label = _label(domain).replace("-", "")
    words = [w for w in re.findall(r"[a-z0-9]+", company.lower()) if w not in COMPANY_STOPWORDS]
    if any(len(w) >= 3 and w in label for w in words):
        return True
    all_words = re.findall(r"[a-z0-9]+", company.lower())
    acronym = "".join(w[0] for w in all_words if w not in {"pvt", "ltd", "private", "limited"})
    return len(acronym) >= 2 and acronym in label


def find_company_domain_mismatch(
    company_name: str | None, text: str, sender_emails: list[str]
) -> list[str]:
    """Sender domains that don't belong to the company the letter claims to be from.

    Known brands the letter claims to be from are checked against their
    official domains; any other company name (from Gemini) is matched loosely.
    """
    sender_domains = list(dict.fromkeys(
        registrable_domain(email_domain(e)) or email_domain(e) for e in sender_emails
    ))
    if not sender_domains:
        return []
    hits: list[str] = []
    brands = claimed_brands(company_name, text)
    for brand in brands:
        if not any(is_official_domain(d, brand) for d in sender_domains):
            hits.append(
                f"Letter mentions {brand.name} but emails come from {', '.join(sender_domains)}"
            )
    if company_name and not brands and not any(_name_matches_domain(company_name, d) for d in sender_domains):
        hits.append(f"Company '{company_name}' does not match sender domain(s) {', '.join(sender_domains)}")
    return hits


def find_suspicious_tlds(domains: list[str]) -> list[str]:
    """Domains on TLDs heavily abused for phishing."""
    return [f"{d} (.{tld_of(d)})" for d in domains if tld_of(d) in SUSPICIOUS_TLDS]


def find_shorteners(urls: list[str]) -> list[str]:
    """URLs that go through a link shortener."""
    return [u for u in urls if (host_of(u) or "") in URL_SHORTENERS]


def find_insecure_urls(urls: list[str]) -> list[str]:
    """URLs explicitly using plain ``http://``."""
    return [u for u in urls if u.lower().startswith("http://")]
